Vitally organizations

anathomical opened the pull request:
## WHAT
- Sync data to Vitally when `SalesFormSubmission` records get created
- Allow Intercom to send sales opportunities to our system in the form of `SalesFormSubmission` records
- Sync `District` records to Vitally as `Organizations`
## WHY
This should allow us to do almost all customer journey management in Vitally
## HOW
CISSY MAGIC (that is to say, Cissy wrote all the code and most of the tests before she left, I finished about a dozen `NotImplemented` tests).
- Modify the `SalesFormSubmission` model to do a bunch of Vitally API calls `after_save` so that those opportunities sync to Vitally
- Add a new webhook controller that accepts data from Intercom, validates that it's cryptographically signed with our shared key, and then creates `SalesFormSubmission` records based on the payload
- Add logic to the `SyncVitallyWorker` to figure out which `Districts` we should sync to Vitally and then enqueue those `Districts` for upload to Vitally via the `SyncVitallyOrganizations` worker.  (We call them Districts, and Vitally calls them Organizations.)

### Notion Card Links
https://www.notion.so/quill/Move-the-sales-opportunity-pipeline-from-Streak-to-Vitally-project-8f04ef91406844c5a3746add11304e0c

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

cissyxyu committed:
all initial files

cissyxyu committed:
initial test files'

cissyxyu committed:
add some limits bc sidekiq is too slow

cissyxyu committed:
sync vitally no limit

cissyxyu committed:
Merge branch 'develop' of https://github.com/empirical-org/Empirical-Core into vitally-organizations

cissyxyu committed:
add opportunities flow

cissyxyu committed:
take out unnecessary serializers

cissyxyu committed:
add all spec files

cissyxyu committed:
linting

cissyxyu committed:
linting again

cissyxyu committed:
create new schools

cissyxyu committed:
testing out intercom controller

cissyxyu committed:
fixes to get intercom hooks endpoint working

cissyxyu committed:
remove unnecessarsy binding

cissyxyu committed:
add uncreated code part with comment

cissyxyu committed:
get user role changes working

cissyxyu committed:
intercom changes

cissyxyu committed:
fix parsed payload

cissyxyu committed:
fix another typo

cissyxyu committed:
change payload parsing to accept contact as well as user

cissyxyu committed:
change wufoo links to live links

cissyxyu committed:
find by instead of find

cissyxyu committed:
try anonymou false check

cissyxyu committed:
pull from develop

cissyxyu committed:
add school

cissyxyu committed:
merge conflict

cissyxyu committed:
change validations to allow nulls

cissyxyu committed:
account for when leads are sent over with no name

cissyxyu committed:
small fixes requested by peter sharkey to district manager

cissyxyu committed:
linting fixes

cissyxyu committed:
more linting

cissyxyu committed:
add spec for thomas

cissyxyu committed:
fix unneeded change

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into vitally-organizations

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into vitally-organizations

anathomical committed:
Add tests for intercome webhooks controller

anathomical committed:
Update cases for accounts controller spec

anathomical committed:
Update districts controller spec with missing cases

anathomical committed:
Add test cases for exists? behavior in the REST API

anathomical committed:
Fix linting error

anathomical committed:
Clean up tests on SalesFormSubmission

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into vitally-organizations

anathomical committed:
Fix linting and clean up an old test

anathomical committed:
Make sure that we add the new user-type to specs

brendanshean commented on services/QuillLMS/app/services/vitally_integration/vitally_rest_api.rb:
```diff
@@ -0,0 +1,46 @@
+
+  def get(type, id)
+    HTTParty.get("#{VITALLY_REST_API_BASE_URL}/#{type}/#{id}",
+      headers: {
```
I would refactor headers for the three methods

```
private def headers
   {  Authorization: "Basic #{@api_key}",  "Content-Type": "application/json" }
end
```

brendanshean commented on services/QuillLMS/app/views/pages/home_new.html.erb:
```diff
@@ -415,7 +415,7 @@ Quill Grammar has over 150 sentence writing activities to help your students. Ou
     <p class="premium-text">Quill Premium's advanced reporting features are the best way to support teachers at the school or district level.</p>
     <div class="premium-button-container">
-      <a class="quill-button medium primary contained sales-button" href="https://quillpremium.wufoo.com/forms/quill-premium-quote/" rel="noopener noreferrer" target="_blank">Contact sales</a>
+      <a class="quill-button medium primary contained sales-button" href="https://quill.org/request_quote" rel="noopener noreferrer" target="_blank">Contact sales</a>
```
I would use path helpers here instead of hardcoded urls:

```
      <a class="quill-button medium primary contained sales-button" href=<%= request_quote_path %>  noreferrer" target="_blank">Contact sales</a>
      <a class="quill-button medium secondary outlined" href=<%= premium_path %> target="_blank">Learn more</a>
```

brendanshean commented on services/QuillLMS/app/services/vitally_integration/vitally_rest_api.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def exists?(type, id)
+    get(type, id).try(:error).blank?
```
I tend to use safe navigation over `try` due to [performance](https://jetrockets.com/blog/nswsheigth-safe-navigation-vs-try-in-rails-part-2-performance).

I realize they are not quite equivalent but just wondering.

anathomical replied:
So it turned out that the `try` was attempting to cover an edge case where `HTTParty` might sometimes have access to an `error` attribute and sometimes not.  But it turns out that's not how `HTTParty` works, so I ended up reworking this line.

brendanshean replied:
Thank you for doing a deeper investigation.

brendanshean commented on services/QuillLMS/app/services/vitally_integration/vitally_rest_api.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def create(type, payload)
+    post(type, payload)
```
what about removing the private post method and just moving it into create?

brendanshean commented on services/QuillLMS/app/workers/sync_vitally_worker.rb:
```diff
@@ -28,10 +31,14 @@ def perform
 
   def schools_to_sync
-    School.select(:id).distinct.joins(:users).where('users.role = ?', 'teacher')
+    @schools_to_sync ||= School.select(:id, :district_id).distinct.joins(:users).where('users.role = ?', 'teacher').where.not(district_id: nil)
```
Do we want the `where.not(district_id: nil)`?  That makes sense for the district case, but  `.compact` will handle that case in the `districts_to_sync`.

anathomical replied:
I didn't write that addition, but based on my review of the specs, I think that you're right.  I'll remove the condition.

brendanshean commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/current_subscription.jsx:
```diff
@@ -40,7 +40,7 @@ export default class CurrentSubscription extends React.Component {
     }
     if (payment_method !== CREDIT_CARD && (authorityLevel || !purchaser_email)) {
-      return <span>{baseText} To renew your subscription for next year, <a href="mailto:sales@quill.org">contact us</a>.</span>
+      return <span>{baseText} To renew your subscription for next year, <a href="https://quill.org/request_renewal">contact us</a>.</span>
```
I would extract `https://quill.org/request_renewal` into a prop and specify `requestQuotePath: request_quote_path` [here](https://github.com/empirical-org/Empirical-Core/blob/fd7c4a8a9d5840e61aba12273337b108475adcdc/services/QuillLMS/app/views/pages/shared/_premium_main.html.erb#L22)

eadams17 replied:
we'd want `rel="noopener noreferrer" target="_blank"` for this link

brendanshean commented on services/QuillLMS/client/app/bundles/Teacher/components/modules/questionsAndAnswers/premium.jsx:
```diff
@@ -15,7 +15,7 @@ const premium = () => [
   {
     question: 'How much does premium cost?',
-    answer: <p>Quill Premium is provided as a teacher license, school license, or district license. The teacher license covers all of the students for one teacher and costs $80 per year. The school license is a site-wide license that covers all teachers and students in the school. For districts, we provide custom pricing, on-site training, and district dashboards. <a href="https://quillpremium.wu ...
+    answer: <p>Quill Premium is provided as a teacher license, school license, or district license. The teacher license covers all of the students for one teacher and costs $80 per year. The school license is a site-wide license that covers all teachers and students in the school. For districts, we provide custom pricing, on-site training, and district dashboards. <a href="https://quill.org/reque ...
```
I would extract `https://quill.org/request_quote` into a prop and specify `requestQuotePath: request_quote_path` [here](https://github.com/empirical-org/Empirical-Core/blob/fd7c4a8a9d5840e61aba12273337b108475adcdc/services/QuillLMS/app/views/pages/shared/_premium_main.html.erb#L22)

brendanshean commented on services/QuillLMS/client/app/bundles/Teacher/components/premium/premium_minis/school_pricing_mini.jsx:
```diff
@@ -16,6 +16,7 @@ const SchoolPricingMini = ({ plan, premiumFeatureData, showBadges, handleClickPu
       <div className="premium-button-container">
         <button
           className="quill-button contained medium primary focus-on-light"
+          href="https://quill.org/request_quote"
```
I would use the prop specified earlier.

brendanshean commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -176,6 +176,26 @@ class User < ApplicationRecord
 
   validate :validate_flags
 
+  TEACHER = 'teacher'
```
There's some duplication here with constants.  They are defined higher up in the file.

brendanshean commented on services/QuillLMS/app/controllers/cms/districts_controller.rb:
```diff
@@ -48,10 +48,13 @@ def edit_subscription
   def create
     new_district = District.new(district_params)
-    if new_district.save
+    begin
```
I feel like begin/rescue exception handling is being [overused](https://dalibornasevic.com/posts/52-don-t-overuse-exceptions) here.  Some if/else conditionals would probably suffice.

anathomical replied:
Yeah, that makes sense to me.  I'll update.

brendanshean commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,94 @@
+    end
+
+    private def find_or_create_user
+      user = User.find_by(id: user_payload["user_id"]) if !user_payload["anonymous"]
```
`unless` might be easier to read.

brendanshean commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,94 @@
+      hexdigest = OpenSSL::HMAC.hexdigest('sha1', client_secret, payload)
+      return if request.headers['X-Hub-Signature'] == "sha1=#{hexdigest}"
+
+      raise "unauthorized call of Intercom webhook"
```
Should this be extracted into an Error class?

anathomical replied:
Yeah, that seems smart.

brendanshean commented on services/QuillLMS/spec/workers/sync_vitally_worker_spec.rb:
```diff
@@ -8,12 +8,39 @@
   describe "#perform" do
     it "make queries for schools and users and enqueue them for further jobs" do
       ENV['SYNC_TO_VITALLY'] = 'true'
```
All the `ENV['SYNC_TO_VITALLY']` calls in these tests should be removed since they can leak state.
I would opt a single  `before { stub_const('ENV', {'SYNC_TO_VITALLY' => 'true'}) }` that applies to all of the specs instead.

brendanshean reviewed (commented):
I've left some comments, but I'm still only halfway through the PR.

anathomical committed:
Make changes based on PR feedback

anathomical committed:
Small tweak to api.exists? call

anathomical commented:
@brendanshean I think I've addressed all of the stuff you called out in you pass so far.

anathomical committed:
Clean up duplicate constant definitions

anathomical committed:
Fix linting

brendanshean commented:
> @brendanshean I think I've addressed all of the stuff you called out in you pass so far.

Thank you @anathomical.  Looks great.  I will take a look at the rest of the PR.

brendanshean commented on services/QuillLMS/app/models/district.rb:
```diff
@@ -27,6 +27,7 @@ class District < ApplicationRecord
   include Subscriber
 
   validates :name, presence: true
+  validates :nces_id, uniqueness: true
```
Good call on validation here.

brendanshean commented on services/QuillLMS/spec/controllers/sales_form_submission_controller_spec.rb:
```diff
@@ -5,9 +5,21 @@
-    let!(:sales_form_submission) { create(:sales_form_submission) }
+    let!(:school) { create(:school) }
+
+    before do
```
Are these three stubs necessary?

brendanshean commented on services/QuillLMS/spec/controllers/sales_form_submission_controller_spec.rb:
```diff
@@ -5,9 +5,21 @@
+      sales_form_submission = double
+      allow(SalesFormSubmission).to receive(:new).and_return(sales_form_submission)
+      allow(sales_form_submission).to receive(:save!).and_return(true)
+      allow(sales_form_submission).to receive(:send_opportunituy_to_vitally)
```
Looks like there's a typo `:send_opportunituy_to_vitally`

brendanshean commented on services/QuillLMS/spec/controllers/sales_form_submission_controller_spec.rb:
```diff
@@ -5,9 +5,21 @@
+      allow(sales_form_submission).to receive(:save!).and_return(true)
+      allow(sales_form_submission).to receive(:send_opportunituy_to_vitally)
+
+      expect(sales_form_submission).to receive(:save!)
```
Were lines 19 and 22 what we discussed at the last engineering meeting?

brendanshean commented on services/QuillLMS/spec/controllers/intercom_integration/webhooks_controller_spec.rb:
```diff
@@ -0,0 +1,96 @@
+      it 'raises error' do
+        expect { post :create }
+          .to raise_error("unauthorized call of Intercom webhook")
+      end
```
I would use the exception class that you have now defined:

`it { expect { post :create }.to raise_error(described_class::UnauthorizedIntercomWebhookCallError) }`

brendanshean commented on services/QuillLMS/app/models/sales_form_submission.rb:
```diff
@@ -30,18 +32,192 @@ class SalesFormSubmission < ApplicationRecord
   validates :collection_type, presence: true, inclusion: { in: COLLECTION_TYPES }
   validates :submission_type, presence: true, inclusion: { in: SUBMISSION_TYPES }
+
+  after_save :vitally_callbacks
```
I tend to avoid after_save callbacks except those managing state internal to the object.  (e.g. see this [post](https://samuelmullen.com/2013/05/the-problem-with-rails-callbacks/)).  Maybe discuss offline pros/cons?

anathomical replied:
For posterity: Brendan and I had a call about this and basically decided to compromise.  We want the `after_save` because we want to guarantee that this data syncs to Vitally no matter how we end up creating these records.  However, we also definitely don't want to tightly couple the ability to save models to external APIs (which is what happens if you do this synchronously).  So we're keeping `after_save`, but moving the sync work into a Sidekiq worker that's triggered by said `after_save` hook.

brendanshean reviewed (commented):
Some more comments

anathomical committed:
Refactor syncing data to Vitally so it happens in a Sidekiq job

So that our general models .save function isn't dependent on an external api

anathomical committed:
Fix a couple of minor items caught in code review

eadams17 commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/current_subscription.jsx:
```diff
@@ -72,7 +72,7 @@ export default class CurrentSubscription extends React.Component {
         To renew your subscription for next year, contact us now
-        at <a href="mailto:sales@quill.org">sales@quill.org</a>.
+        at <a href="https://quill.org/request_renewal">sales@quill.org</a>.
       </span>
```
I think we'd probably want to change this to maybe have "contact us now" be the displayed link text (and also add `rel="noopener noreferrer" target="_blank"`)

eadams17 reviewed (approved):
looks good! just a few small comments around links

happythenewsad commented on services/QuillLMS/db/migrate/20220609175032_change_sales_form_submission_columns.rb:
```diff
@@ -0,0 +1,10 @@
+    change_column_null :sales_form_submissions, :phone_number, true
+    change_column_null :sales_form_submissions, :zipcode, true
+    change_column_null :sales_form_submissions, :school_name, true
+    change_column_null :sales_form_submissions, :district_name, true
```
school_name and district_name don't have the non-null constraint, so lines 7,8 should be removed because they are redundant.

happythenewsad commented on services/QuillLMS/spec/models/sales_form_submission_spec.rb:
```diff
@@ -11,48 +11,77 @@
-    it { expect(sales_form_submission).to be_valid }
-    it { should validate_presence_of(:school_name) }
-    it { should validate_presence_of(:district_name) }
+    #before do
```
remove comments

happythenewsad commented on services/QuillLMS/spec/models/user_spec.rb:
```diff
@@ -346,8 +346,8 @@
 
   describe 'constants' do
     it "should give the correct value for all the contstants" do
```
typo

brendanshean commented on services/QuillLMS/spec/models/sales_form_submission_spec.rb:
```diff
@@ -11,48 +11,77 @@
+      expect(SyncSalesFormSubmissionToVitallyWorker).to receive(:perform_async).with(sales_form_submission.id)
+      expect(SendSalesFormSubmissionToSlackWorker).to receive(:perform_async).with(sales_form_submission.id)
+
+      sales_form_submission.created_at = Time.now
```
I'm not sure why the linter didn't catch this but we should use Time.current over Time.now.   Also, what's the reasoning behind updating the created_at here.

brendanshean reviewed (approved):
I think the refactoring after_save to use workers looks much better.  Thanks for finishing the last mile of this large PR.

dandrabik commented on services/QuillLMS/app/controllers/cms/districts_controller.rb:
```diff
@@ -48,10 +48,11 @@ def edit_subscription
   def create
     new_district = District.new(district_params)
-    if new_district.save
+    if new_district.valid?
```
I think this was actually better as `if new_district.save` vs. `.valid?` and then `save!`

dandrabik commented on services/QuillLMS/app/controllers/cms/districts_controller.rb:
```diff
@@ -48,10 +48,11 @@ def edit_subscription
       redirect_to cms_district_path(new_district)
     else
-      redirect_to cms_districts_path, error: new_district.errors, flash: { error: new_district.errors }
+      redirect_to cms_districts_path, flash: { error: "A district with this NCES ID already exists." }
```
Are we sure this is what the error would be? Seems like the previous approach gave more specific errors.

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+    before_action :verify_signature, only: [:create]
+
+    def create
+      topic = parsed_payload["topic"]
```
Is there a reason we're not using standard Rails `params` here?

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+          intercom_link: intercom_link(user_payload["id"])
+        )
+        sales_form_submission.collection_type = collection_type(tag)
+        sales_form_submission.submission_type = submission_type(tag)
```
Could we move these two lines inside the `.new`, . e.g `(collection_type: collection_type(tag), submission_type = submission_type(tag))`

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+      if ["user.tag.created", "contact.tag.created"].include?(topic)
+        tag = parsed_payload["data"]["item"]["tag"]["name"]
+        user = find_or_create_user
+        sales_form_submission = SalesFormSubmission.new(
```
If you add `collection_type` and `submission_type` as params, you can change this to `.create!` and remove the `.save!`. Also, `sales_form_submission` isn't used so you don't need to assign.

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+
+    def create
+      topic = parsed_payload["topic"]
+      if ["user.tag.created", "contact.tag.created"].include?(topic)
```
Maybe make this this a constant to help give context, e.g. `ACCEPTED_WEBHOOK_TOPICS.include?(topic)` or whatever it should be called.

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+      "https://app.intercom.com/a/apps/v2ms5bl3/users/#{intercom_user_id}/all-conversations"
+    end
+
+    private def payload
```
Yeah, let's definitely use Rails `params` instead of this custom code.

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+        email: user_payload["email"],
+        role: User::SALES_CONTACT,
+        name: user_payload["name"] || UNKNOWN_NAME,
+        password: SecureRandom.uuid
```
So we are generating a password for them that they won't know?

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+
+    private def collection_type(tag)
+      case tag
+      when "Quote Request School", "Renewal Request School"
```
Strings like `"Quote Request School"` are used multiple times for logic, let's make them constants.

dandrabik commented on services/QuillLMS/app/controllers/intercom_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,96 @@
+      case tag
+      when "Quote Request School", "Renewal Request School"
+        SalesFormSubmission::SCHOOL_COLLECTION_TYPE
+      when "Quote Request District", "Renewal Request District"
```
For these type of lookups, consider using a hash constant and using `.fetch(key, default)`, e.g.

```ruby
COLLECTION_MAPPING = {
   QUOTE_REQUEST => SalesFormSubmission::DISTRICT_COLLECTION_TYPE,
   RENEWAL_REQUEST => SalesFormSubmission::DISTRICT_COLLECTION_TYPE,
...
}

dandrabik commented on services/QuillLMS/app/services/vitally_integration/vitally_rest_api.rb:
```diff
@@ -0,0 +1,40 @@
+  VITALLY_REST_API_BASE_URL = 'https://rest.vitally.io/resources'
+
+  def initialize
+    @api_key = ENV['VITALLY_REST_API_KEY']
```
This can be a constant and you can drop the `initialize` method. e.g. `API_KEY = ENV['VITALLY_REST_API_KEY']`

dandrabik commented on services/QuillLMS/app/workers/sync_sales_form_submission_to_vitally_worker.rb:
```diff
@@ -0,0 +1,107 @@
+  def create_school_or_district_if_none_exist
+    if (@sales_form_submission.district_collection? &&
+        @sales_form_submission.district.present? &&
+        !api.exists?(SalesFormSubmission::VITALLY_DISTRICTS_TYPE, @sales_form_submission.district.id))
```
Consider a `create_unless_exists` method for the API wrapper, e.g.

```ruby
def create_unless_exists(type, data)
  return if exists?(type, data)

  create(type, data)
end

dandrabik commented on services/QuillLMS/app/workers/sync_sales_form_submission_to_vitally_worker.rb:
```diff
@@ -0,0 +1,107 @@
+  end
+
+  def create_school_or_district_if_none_exist
+    if (@sales_form_submission.district_collection? &&
```
I think the logic here is slightly off and the `api.exists?` shouldn't be in the branching checks.

Code as is:
- If it has district data AND that district IS in vitally, it will try next to create a school
- If it has district data AND that district IS NOT in vitally, it will create a district, but NOT try to create a school

dandrabik commented on services/QuillLMS/app/workers/sync_sales_form_submission_to_vitally_worker.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+  end
+
+  def send_opportunity_to_vitally
```
Should this be a `create_unless_exists`?

dandrabik commented on services/QuillLMS/spec/controllers/accounts_controller_spec.rb:
```diff
@@ -115,6 +115,36 @@
+        let!(:user) { create(:user, role: User::SALES_CONTACT) }
+
+        it 'should use that same user record but update all the fields' do
+          name = "Test Name"
```
Consider making these `let` vars since they are in multiple tests.

dandrabik commented on services/QuillLMS/spec/services/vitally_integration/vitally_rest_api_spec.rb:
```diff
@@ -0,0 +1,90 @@
+  let(:api)  { VitallyRestApi.new }
+
+  before do
+    stub_const('ENV', {'VITALLY_REST_API_KEY' => 'test api key'})
```
When we make API_KEY a class constant, let's stub that instead to avoid touching the global space.

dandrabik commented on services/QuillLMS/spec/services/vitally_integration/vitally_rest_api_spec.rb:
```diff
@@ -0,0 +1,90 @@
+      payload = 'test payload'
+      expect(HTTParty).to receive(:post).with("#{VitallyRestApi::VITALLY_REST_API_BASE_URL}/type",
+        headers: {
+          Authorization: "Basic #{ENV['VITALLY_REST_API_KEY']}",
```
I think you'd want to test that this is `Authorization: "Basic 'test api key'"`, i.e. if you are stubbing it, you should test that the expected stub is coming through.

dandrabik commented on services/QuillLMS/spec/services/vitally_integration/vitally_rest_api_spec.rb:
```diff
@@ -0,0 +1,90 @@
+
+  describe '#get' do
+    it 'should make a GET call to the Vitally API with the specified type and ID' do
+      id = 1
```
Vars that could be reusable `let`s.

dandrabik commented on services/QuillLMS/spec/workers/send_sales_form_submission_to_slack_worker_spec.rb:
```diff
@@ -0,0 +1,21 @@
+      stub_url = 'fake_slack_url'
+      # We only actually try to send messages if RAILS_ENV is production
+      stub_const('ENV', {'SLACK_API_WEBHOOK_SALES' => stub_url,
+                         'RAILS_ENV' => 'production'})
```
Let's avoid changing the `RAILS_ENV` to 'production' for a test. Using a constant in the class that you can stub.

dandrabik commented on services/QuillLMS/spec/controllers/concerns/slack_tasks_spec.rb:
```diff
@@ -5,10 +5,14 @@
+  let(:sales_form_submission) { build(:sales_form_submission) }
 
   before do
     stub_const('ENV', {'RAILS_ENV' => 'production', 'SLACK_API_WEBHOOK_SALES' => 'slack-test.com'})
```
Outside this PR, but let's avoid changing the `RAILS_ENV` to `production`.

Let's add a local constant to SlackTasks and stub that if needed:

```ruby
Class Slack
  LIVE = Rails.env.production?

dandrabik reviewed (changes requested):
Thanks @anathomical for moving this along. I have a bunch of really small clean-up asks in my comments. None of them are particularly large, but since there is a good number of them, I'm marking it as "request changes" and will re-review when prompted.

Thanks for pitching and helping to carry this code along.

anathomical committed:
Update files based on code review

anathomical committed:
Do a bunch of clean up based on code review

anathomical committed:
Final tweaks to tests to hopefully resolve linting errors

anathomical committed:
Fix linting error

anathomical committed:
Test validator with custom message

anathomical committed:
Update a snapshot for the front-end

dandrabik reviewed (approved):
:+1: Thanks for making all of these changes and pushing this code forward.

happythenewsad reviewed (approved):
LGTM

anathomical committed:
Make some tweaks for safety around 'sales-contact' User roles

We need to do some more thinking about how we're going to handle that role and exactly how we want it to work, so flag all the parts of the code that we're merging to production that won't actually be in use immediately so that it's easy to find them once we resolve those issues

anathomical committed:
A minor bit of clean-up about our safety disabling stuff

anathomical committed:
Fix linting errors

anathomical committed:
Fix another linting erro

anathomical committed:
Another linting pass

anathomical committed:
Linting.  Sigh
